enhancements/machine-config/mco-cgroupsv2-support.md

```diff
@@ -0,0 +1,192 @@
+all workloads will be compatible with cgroup v2, so it will *not* be enabled by
+default within OpenShift at this time.
+
+## Motivation
+
+Migrating to cgroup v2 will bring in many new features and fixes not found in
+cgroup v1. cgroup v1 is considered 'legacy' and migrating to cgroup v2 is
+considered necessary since RHEL ships with cgroup v2 on by default. (OpenShift
+4.x currently disables cgroup v2 in favor of v1).
+
+Some features of cgroup v2 include:
+
+* IO enhancements
+* User based OOM killer
+* cgroup aware OOM killer
+
+[Kubernetes On Cgroup v2 - Video](https://www.youtube.com/watch?v=u8h0e84HxcE)
+
+### Goals
+
+- [ ] Enable cgroup v2 within the Openshift API
+- [ ] Add kernel flags to MCO to enable cgroup v2 on nodes
+
+### Non-Goals
+
+- Support mixed cgroups modes
+
+## Proposal
+
+The option to enable cgroup v2 will have to reside in a centralized location.
+The [OpenShift Infrastructure config
+object](https://github.com/openshift/api/blob/master/config/v1/types_infrastructure.go#L28)
+contains information describing how a cluster functions including cloud config  
+and platform specification for each cloud. Setting the cgroup mode is an
+infrastructure setting.
+
+Create an additional spec within `Infrastructure` named `cgroupSpec` to contain
+which version of cgroups to enable.
+
+```diff
+diff --git a/config/v1/types_infrastructure.go b/config/v1/types_infrastructure.go
+index 4863a98e..4b2bd2b1 100644
+--- a/config/v1/types_infrastructure.go
++++ b/config/v1/types_infrastructure.go
+@@ -2,6 +2,13 @@ package v1
+ 
+ import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+ 
++type CgroupMode string
++
++const (
++	CgroupMode_v1 = "v1"
++	CgroupMode_v2 = "v2"
++)
++
+ // +genclient
+ // +genclient:nonNamespaced
+ // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+@@ -45,6 +52,13 @@ type InfrastructureSpec struct {
+ 	// platformSpec holds desired information specific to the underlying
+ 	// infrastructure provider.
+ 	PlatformSpec PlatformSpec `json:"platformSpec,omitempty"`
++
++	// cgroupSpec holds the desired cgroup version to use for a cluster
++	CgroupSpec CgroupSpec `json:"cgroupSpec,omitempty"`
++}
++
++type CgroupSpec struct {
++	Mode CgroupMode `json:"mode"`
+ }
+ 
+ // InfrastructureStatus describes the infrastructure the cluster is leveraging.
+```
+
+Once the previous API is defined, the MCO will read the infrastructure option
+(via bootstrap or day2 operations).
+
+The following kernel command line arguments would be set when `CgroupMode_v2` is enabled:
+```yaml
+  kernelArguments:
+    - systemd.unified_cgroup_hierarchy=1
+    - cgroup_no_v1="all"
+    - psi=1 
+```
+
+### User Stories
+
+#### 1. Modify the Openshift API with the agreed upon spec
+
+#### 2. Modify the MCO to accept the new Openshift API and enable cgroup v2 kernel options (bootstrap)
+
+#### 3. Modify the MCO to accept the new Openshift API and enable cgroup v2 kernel options (day 2)
+
+### Risks and Mitigations
```

Comment:
Although this is an implementation detail for MCO, but with adding cgroup option MCO will need to ensure that user is not also adding any cgroup v2 specific kargs via a MachineConfig. Otherwise, I think this can lead to undesired behavior when cluster infraSpec has cgroup v1.